File: backend/api/conversations/src/main/java/co/airy/core/api/conversations/dto/Conversation.java
package co.airy.core.api.conversations.dto;

import co.airy.avro.communication.Channel;
import co.airy.avro.communication.Message;
import co.airy.avro.communication.MetadataKeys;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Conversation implements Serializable {
    Long createdAt;
    String channelId;
    Message lastMessage;
    String sourceConversationId;
    Channel channel;

    Integer unreadCount;

    @Builder.Default
    Map<String, String> metadata = new HashMap<>();

    public String getDisplayName() {
        final String firstName = this.metadata.get(MetadataKeys.SOURCE.CONTACT.FIRST_NAME);
        final String lastName = this.metadata.get(MetadataKeys.SOURCE.CONTACT.LAST_NAME);

        if (firstName == null && lastName == null) {
            return null;
        }

        return String.format("%s %s", firstName, lastName).trim();
    }

    public String getId() {
        return this.lastMessage.getConversationId();
    }

    public String getChannelId() {
        return this.lastMessage.getChannelId();
    }
}

File: lib/typescript/httpclient/model/Conversation.ts
import {Contact} from './Contact';
import {Message} from './Message';
import {Metadata} from './Metadata';
import {Channel} from './Channel';

export type ConversationMetadata = Metadata & {
  contact: Contact;
  unreadCount: number;
  tags: {
    [tagId: string]: string;
  };
};

export interface Conversation {
  id: string;
  channel: Channel;
  metadata: ConversationMetadata;
  createdAt: Date;
  lastMessage: Message;
}

export function getTags(conversation: Conversation) {
  return Object.keys(conversation.metadata.tags || {});
}

export function getSource(conversation: Conversation) {
  return conversation?.channel?.source;
}
